### PR TITLE
[KB][CDS-85737]: Automate your workflows: Using Harness Approval APIs

### PR DESCRIPTION
**Pull request by RishabhGupta34**
# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the CODEOWNERS. 

## What Type of PR is This?

- [ ] Issue
- [ ] Feature
- [ ] Maintenance/Chore

If tied to an Issue, list the Issue(s) here:

* *Issue(s)*

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [ ] Tested Locally
- [ ] *Optional* Screen Shoot.

### CODE REVIEW

File: kb/continuous-delivery/articles/harness-approval-api.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+### A Symphony of Approvals
+
+Consider this sample scenario: Your organization has multiple pipelines, each representing a crucial step in your software deployment process. These pipelines await approvals before progressing further, ensuring a robust and controlled release.
+
+Now, imagine one of these pipelines is waiting for an approval. But, this approval can only be approved after the completion of a process running in an external 3rd party system. The approver, responsible for approving pipelines, would typically:
+
+1. **Navigate the Pipeline Executions:** First, they have to go through the list of pipeline executions to identify the executions waiting for an approval. 
+
+2. **Connect with External Systems:** Upon identifying the pending approval, the approver is then required to switch to the external 3rd party system. Here, they manually interact with the system to grant approval, possibly after validating external conditions or processes.
```
Comment (SudarshanPurohit) [inline kb/continuous-delivery/articles/harness-approval-api.md:24]: What does this mean? We're just saying that "The Approver verifies whether the external system's task is complete and successful? After this is done, she comes back to the Pipeline Execution page to approve it".

File: kb/continuous-delivery/articles/harness-approval-api.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+3. **Repeat the Process:** Generally the approver wouldn't have just one pipeline to approve, but several such pending approvals. Each approval necessitates the same tedious navigation, manual intervention, and decision-making process.
+
+This cumbersome journey not only consumes valuable time but also introduces the potential for errors and delays. As the number of pending approvals increases, so does the complexity of managing and ensuring timely approvals.
```
Comment (SudarshanPurohit) [inline kb/continuous-delivery/articles/harness-approval-api.md:28]: Shorten this, or remove it. Our users know it's cumbersome

File: kb/continuous-delivery/articles/harness-approval-api.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+## Introducing the Approval API: Approve or Reject an Execution by Pipeline Execution ID
+
+This Approval API introduces a seamless method to approve or reject a pipeline by directly using the pipeline's execution ID. It only needs an execution id to approve any pipeline waiting for an approval. So any process can call this API to approve or reject the execution based on the process's status, eliminating the need for manual intervention and making approvals a seamless process.
```
Comment (SudarshanPurohit) [inline kb/continuous-delivery/articles/harness-approval-api.md:32]: Can we add a link to the actual API here, from API Docs? 

Also, I don't think we are "introducing" the new API, we've just made it simpler to use.

File: kb/continuous-delivery/articles/harness-approval-api.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+Approvals play a crucial role in getting your deployments done timely and with precision. Although too many approvals or too many dependencies on external systems can create bottlenecks and increase the complexity of managing the approvals.
+
+Harness understands the importance of this coordination, and that's why we bring you the Approval API which can approve any pipeline by the pipeline execution id. These APIs are designed to simplify and streamline the approval process in your projects, offering flexibility and choice to suit your unique needs.
+
+## The Canvas of Approval
+
+Traditionally, the Harness UI has been the go-to for approving or rejecting pipeline deployments. However, in the era of automation, we present Approval APIs that open up new dimensions for orchestrating your deployments.
+
+### A Symphony of Approvals
```
Comment (SudarshanPurohit) [inline kb/continuous-delivery/articles/harness-approval-api.md:16]: "Canvas", "Symphony", etc. doesn't convey any meaning to the user who quickly wants to solve his use case. I think we should organize this more like a tech doc. 

- What is the use case? 
- What are sample scenarios where Approval is used?
- How is it solved using the UI/Manually?
- How can we solve it better using the API directly? 
- What are the two core use cases we are supporting in the API? 
- What do we do if there's just one Approval? (with an API example)
- What do we do if there are multiple Approvals? (with an example API call, and a screenshot of where to set the approval Id in the UI, and a sample YAML format for Approval step)
- What are the possible errors that the user could get when using the API? What is the solution?
- What use cases are not covered? Is there a workaround?
- What are further links that are relevant?

File: kb/continuous-delivery/articles/harness-approval-api.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+Replace the <step_group_id> by the identifier of the stepgroup when it gets used in a pipeline
+
+:::
+
+
+## Ensuring Single Approval Instance
+
+To ensure clarity and precision, this API is designed to handle only one approval instance at a time. If multiple approvals are pending simultaneously, the API call will encounter an error, signaling that approving multiple instances in a single call is not supported. However, in the scenario where only one approval is awaiting consideration, it will seamlessly proceed to be automatically approved.
+
+## The Power of Choice Unleashed
```
Comment (SudarshanPurohit) [inline kb/continuous-delivery/articles/harness-approval-api.md:195]: What does this API have to do with "power of choice unleashed"? Please, let's keep things simple and clear.

File: kb/continuous-delivery/articles/harness-approval-api.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+- **Streamlined Workflow:** Automate the approval process directly from your systems.
+- **Real-time Integration:** Sync external processes seamlessly with your deployment pipelines.
+- **Efficiency Redefined:** No need for manual interventions—let the Approval API do the heavy lifting.
+
+## Navigating the Future
```
Comment (SudarshanPurohit) [inline kb/continuous-delivery/articles/harness-approval-api.md:201]: We already explained the use case and the API. Let's remove this section

File: kb/continuous-delivery/articles/harness-approval-api.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+Bid farewell to manual time-taking approvals, which could be time consuming and delay your release cycles. With this Approval API, you can automate the approvals, which could have been automated in the first place for e.g. approvals which just require a go ahead from other processes/pipelines.
+
+### Harness Documentation
```
Comment (SudarshanPurohit) [inline kb/continuous-delivery/articles/harness-approval-api.md:205]: This should be referenced everywhere we talk of the API, not down at the bottom.

File: kb/continuous-delivery/articles/harness-approval-api.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
+---
+description: KB - Gain a deep understanding of Harness Approval APIs and how to use them to automate workflows. 
+title: "Orchestrating Deployments with Choice: Exploring Harness Approval APIs"
+---
+
+# Orchestrating Deployments with Choice: Exploring Harness Approval APIs
+
+Approvals play a crucial role in getting your deployments done timely and with precision. Although too many approvals or too many dependencies on external systems can create bottlenecks and increase the complexity of managing the approvals.
+
+Harness understands the importance of this coordination, and that's why we bring you the Approval API which can approve any pipeline by the pipeline execution id. These APIs are designed to simplify and streamline the approval process in your projects, offering flexibility and choice to suit your unique needs.
```
Comment (SudarshanPurohit) [inline kb/continuous-delivery/articles/harness-approval-api.md:10]: Note that the API works along with the Approval step. The API doesn't "simplify and streamline the approval process", it only approves the steps already created and which are pending in the pipeline. Let's not hype this up too much :).

**Review by SudarshanPurohit — COMMENTED** [review]
Let's make this more tech-doc-y and less linkedin-post-y, please. We can remove most of the adjectives and adverbs too.

File: kb/continuous-delivery/articles/harness-approval-api.md (hunk 1 above)
Comment (RishabhGupta34) [reply kb/continuous-delivery/articles/harness-approval-api.md]: Simplified the wording

File: kb/continuous-delivery/articles/harness-approval-api.md (hunk 2 above)
Comment (RishabhGupta34) [reply kb/continuous-delivery/articles/harness-approval-api.md]: Shortened it

File: kb/continuous-delivery/articles/harness-approval-api.md (hunk 3 above)
Comment (RishabhGupta34) [reply kb/continuous-delivery/articles/harness-approval-api.md]: Changed the wording, for api docs link waiting for the API doc generation, will link

File: kb/continuous-delivery/articles/harness-approval-api.md (hunk 4 above)
Comment (RishabhGupta34) [reply kb/continuous-delivery/articles/harness-approval-api.md]: Made it more organized and simplified the language

File: kb/continuous-delivery/articles/harness-approval-api.md (hunk 5 above)
Comment (RishabhGupta34) [reply kb/continuous-delivery/articles/harness-approval-api.md]: removed

File: kb/continuous-delivery/articles/harness-approval-api.md (hunk 7 above)
Comment (RishabhGupta34) [reply kb/continuous-delivery/articles/harness-approval-api.md]: done

File: kb/continuous-delivery/articles/harness-approval-api.md (hunk 6 above)
Comment (RishabhGupta34) [reply kb/continuous-delivery/articles/harness-approval-api.md]: removed

**Review by puthrayaharness — APPROVED** [review]

**Pull request merged**